text/0002-signature-based-completions.md

```diff
@@ -0,0 +1,216 @@
+  - Flag name: `-<f>` and `--<flag>`
+  - Flag value `-f <val>` and `--flag=<val>`
+  - Positional arg: `command --flag -s <arg>`
+  - Math operator: `= value <op> value`
+  - Variable: `$<var>`, `$nu.<var>`
+  - Column path: `thing.<path>`
+
+All of these are found in the context of a pipeline, which is just a series of commands connected by
+`|`. Pipelines can be found in many locations:
+
+  - `pipeline ; pipeline`
+  - `where { pipeline }`
+  - `echo $(pipeline)`
+  - `` `format string {{$(pipeline)}}` ``
+
+The completion engine should understand the context it is in, and this is easiest if we defer most
+of the work to the parser. The parser should be able to inform us about the location we're in. If
+it's the position for a value, it should even describe the shape needed for the value. For example,
+`where` takes one required positional argument, which has a "math" shape. The completion engine can
+then look for a completer or set of completers that understand how to complete a math shape, and
+collect the suggestions (if any).
+
+Being able to write a completer that only cares about a specific shape is a powerful concept. It
+takes away the need from writing completers that need to understand the entire line (although we can
+still allow them to request the full line).
+
+# Reference-level explanation
+
+The major technical portion of this PR will be to overhaul the parser, so that it can better hint at
+the shape under the cursor. The parser currently is two-phased: lite parse and full parse. The lite
+parse does the basics of pulling out some basic components on a line, but knows nothing about
+signatures. If it fails to parse, an error is returned. We need to extend this error with
+information about what was successfully parsed, because the error may have occurred beyond the
+cursor location. For example:
+
+    > ls "some dir/
+
+doesn't parse successfully, since the string is unterminated. That being said, the parser knows that
+it's attempting to parse a string, so it should be able to convey that to the caller so that they
+can take better action on this error state.
+
+After lite parse, we need to follow up with a full parse to figure out which shape is under the
+cursor. If this succeeds, we have a shape. If it errors, we provide no completion suggestions.
+
+The second technical aspect of this completion engine is to make it store a registry of completers.
+This registry would map completion locations and syntax shapes to a list of completers. For example,
+we may have a registry like this:
+
+  | completion location | syntax shape | completers |
+  |---------------------|--------------|------------|
+  | command             | string       | filename, internal command, external command |
+  | flag/arg value      | string       | filename |
+  | flag/arg value      | path         | filename |
+  | flag name           | string       | flag name |
+  | variable            | string       | variable name |
+
```

Comment:
Will we allow users to contribute completion files and have nu consume them for external commands/tools?

Reply:
It's outside the scope of this PR, but definitely, would want that 💯